### ISSUE TITLE
Installation from source: error: #error "AsmJit - Unable to detect host architecture."

### ISSUE THREAD

**Issue by fertinaz** [pandegroup/openmm#2030]
Hi there

Trying to install OpenMM on a GPU cluster (without root) using gcc-7.2.1 and cuda-8.0. OS is Ubuntu 16.04.

I have made following modifications to the CMakeCache.txt
Added: 
```
OPENMM_BUILD_CUDA_LIB:BOOL=ON
```
Removed - otherwise receiving Doxygen error: 
```
OPENMM_BUILD_PYTHON_WRAPPERS=OFF
OPENMM_GENERATE_API_DOCS=OFF
OPENMM_BUILD_C_AND_FORTRAN_WRAPPERS=OFF
```

I removed the "-msse2" flag to avoid unrecognised flag error. 

Could generate Makefile using ccmake successfully but cannot run make because of this error:
```
[  2%] Building CXX object CMakeFiles/OpenMM.dir/openmmapi/src/CompiledExpressionSet.cpp.o
In file included from ....../src/openmm/libraries/asmjit/base.h:12:0,
                 from ....../src/openmm/libraries/asmjit/asmjit.h:368,
                 from ....../src/openmm/libraries/lepton/include/lepton/CompiledExpression.h:43,
                 from ....../src/openmm/openmmapi/include/openmm/internal/CompiledExpressionSet.h:35,
                 from ....../src/openmm/openmmapi/src/CompiledExpressionSet.cpp:24:
....../src/openmm/libraries/asmjit/build.h:111:3: error: #error "AsmJit - Unable to detect host architecture."
 # error "AsmJit - Unable to detect host architecture."
   ^
In file included from ....../src/openmm/openmmapi/include/openmm/internal/CompiledExpressionSet.h:35:0,
                 from ....../src/openmm/openmmapi/src/CompiledExpressionSet.cpp:24:
....../src/openmm/libraries/lepton/include/lepton/CompiledExpression.h:105:40: error: 'asmjit::X86Compiler' has not been declared
     void generateSingleArgCall(asmjit::X86Compiler& c, asmjit::X86XmmVar& dest, asmjit::X86XmmVar& arg, double (*function)(double));
                                        ^
....../src/openmm/libraries/lepton/include/lepton/CompiledExpression.h:105:64: error: 'asmjit::X86XmmVar' has not been declared
     void generateSingleArgCall(asmjit::X86Compiler& c, asmjit::X86XmmVar& dest, asmjit::X86XmmVar& arg, double (*function)(double));
                                                                ^
....../src/openmm/libraries/lepton/include/lepton/CompiledExpression.h:105:89: error: 'asmjit::X86XmmVar' has not been declared
     void generateSingleArgCall(asmjit::X86Compiler& c, asmjit::X86XmmVar& dest, asmjit::X86XmmVar& arg, double (*function)(double));
                                                                                         ^
CMakeFiles/OpenMM.dir/build.make:254: recipe for target 'CMakeFiles/OpenMM.dir/openmmapi/src/CompiledExpressionSet.cpp.o' failed
make[2]: *** [CMakeFiles/OpenMM.dir/openmmapi/src/CompiledExpressionSet.cpp.o] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/OpenMM.dir/all' failed
make[1]: *** [CMakeFiles/OpenMM.dir/all] Error 2
Makefile:138: recipe for target 'all' failed
```

Any help is much appreciated

// Fatih

**Comment by peastman** [comment]
What kind of CPUs does the cluster have?  That header is checking for various standard definitions to indicate what kind of processor it's being compiled for (32 or 64 bit, x86 or ARM).  It isn't finding any of them.

**Comment by fertinaz** [comment]
Thanks for quick respond. CPUs are Power8

**Comment by peastman** [comment]
I thought that might be it.  OpenMM isn't currently supported on Power.  In principle it should be possible to port it, but we haven't done that so far.

**Comment by fertinaz** [comment]
Do you have any suggestions for how to start porting it?

**Comment by peastman** [comment]
If you're inclined to try, there are three pieces of code that are architecture specific.  The first is asmjit, which is the library we use to compile expressions for custom forces.  Since it doesn't support Power, we just need to turn off JIT compilation of expressions.  Everything will still work, but custom forces will be slower on the CPU platform.  Since you're using CUDA, that isn't relevant.

This is quite easy to do.  It's controlled by a block in the top level CMakeLists.txt: https://github.com/pandegroup/openmm/blob/master/CMakeLists.txt#L227-L234  Just modify the condition so that block also doesn't get executed on Power.

The second change is to the way we implement vectorization.  This is the file that gets included by any code that uses vector operations: https://github.com/pandegroup/openmm/blob/master/openmmapi/include/openmm/internal/vectorize.h.  It then includes one of a few different files depending on the architecture.  Each one defines the same classes and functions, but implemented with intrinsics for a different instruction set.  We have versions for SSE (x86) and NEON (ARM).  You would need to create a third version that implements them using AltiVec.

Also notice there's a version called vectorize_pnacl.h.  I created it when I was experimenting with Chrome portable native client (PNaCl), but the code isn't actually specific to that.  It implements the operations using the LLVM portable vector API.  It might be fairly easy to get that running on Power.  It's not as fast as a full AltiVec implementation would be, but again that isn't too important if you're using CUDA instead of CPU.

The final piece is the file hardware.h.  It defines a few hardware specific functions, such as querying the number of processors and the processor capability flags.  You might need to make a few changes there.

**Comment by fertinaz** [comment]
Great! Well, thanks very much. I'll try and let you know as well if I manage to succeed.

**Comment by jchodera** [comment]
@fertinaz : Is this the new [Power8 architecture with integrated NVLINK](https://www.ibm.com/blogs/systems/ibm-nvidia-present-nvlink-server-youve-waiting/)?

**Comment by fertinaz** [comment]
Yes, exactly! This is the cluster: https://wiki.scinet.utoronto.ca/wiki/index.php/SOSCIP_GPU

**Comment by jchodera** [comment]
Awesome! 

@peastman: Is OpenMM currently able to take advantage of any of the fast NVLINK interconnects that this new architecture provides?

**Comment by peastman** [comment]
Hopefully, but I haven't looked into it.  Peer-to-peer `cuMemcpy()` should in principle be much faster.

**Comment by mds322** [comment]
@peastman @fertinaz : I am actually trying to do the same thing and notice that there is one other piece of code that seems to attempt to use the SSE2 (after adjusting the cmake file), the random number generator located in:  /openmm/libraries/sfmt/src/SFMT.cpp . And suggestions on what would be the best way around this issue?

**Comment by peastman** [comment]
SFMT can be compiled with various vector instruction sets, or with none at all.  We compile it with or without SSE depending on your platform:

https://github.com/pandegroup/openmm/blob/master/CMakeLists.txt#L222-L226

It also includes AltiVec code, which you can enable by specifying `-DHAVE_ALTIVEC=1`.

**Comment by mds322** [comment]
@peastman seem to have made some progress tinkering with the cmake file, and I  can get make to reach 98% completion and then error out due to the generation of the testvector and testvector8 tests. Cmake is not my forte', how would I go about disabling these tests from being made?

**Comment by peastman** [comment]
I think what you actually want to do is change the compilation options so they compile correctly:

https://github.com/pandegroup/openmm/blob/master/tests/CMakeLists.txt#L18-L20

TestVectorize8 isn't important, since 8 component vectors are only used when AVX is available.  But TestVectorize is an important test to make sure your other changes are working correctly.

**Comment by hengma1001** [comment]
Hi, all. 
I was working on the installation of OpenMM on a PPC with gcc 6.3 and cuda 9.0. 

I got stuck by the vectorize.h of course, and will start working on it later. 
Before that, I noticed there was also another piece of code using sse2 at the library, 

https://github.com/pandegroup/openmm/blob/master/libraries/vecmath/src/vecmath.cpp

handling the vector intrinsics implementation of sin, cos, exp and log. 
Will it be essential to also create PPC version of this library? Is it possible for me to move some cpu code out of the make list that won't influence the performance on GPU later?

**Comment by hengma1001** [comment]
Hi, all. 

I was working compiling OpenMM code on a PPC 8/9 machine with gcc 6.4 and cuda 9.0. 

I got stuck with the vectorize.h, of course. Before that, I noticed sse2 was also used by another piece of code in the libraries, 
https://github.com/pandegroup/openmm/blob/master/libraries/vecmath/src/vecmath.cpp
handling some mathematical functions: sin, cos, log and exp, in vector intrinsics. 
Should it also be converted to ppc or a non-vector version? 

Moreover, how feasible will it be to skip building the some of cpu codes by modifying the make list? Will be the GPU platform affected if I do that? 

Thank you very much.

**Comment by peastman** [comment]
vectorize_sse.h and vectorize_neon.h use the vectorized implementations of exp() and log() from that library: https://github.com/pandegroup/openmm/blob/master/openmmapi/include/openmm/internal/vectorize_sse.h#L40-L42.  vectorize_pnacl.h instead implements them by just calling the regular math function on each element separately: https://github.com/pandegroup/openmm/blob/master/openmmapi/include/openmm/internal/vectorize_pnacl.h#L236-L242.  That's a lot slower, but it still works.  And the speed is only an issue if you plan to use the CPU platform.

**Comment by hengma1001** [comment]
Thank you for your prompt reply. Other than those two, it shouldn't be too difficult to replace all other the operations in the vectorize_sse.h with IBM vmx/vsx/altivec intrinsic. 
One other question is whether the sin() and cos() function would be used elsewhere, such as in a custom angle force, where a trigonometric function might be used?

**Comment by peastman** [comment]
That's the only place the vecmath library gets used.  Custom forces aren't currently vectorized.  That's something I might like to implement some day, but it would probably only be on x86 (since that's all that AsmJit supports).

**Comment by hengma1001** [comment]
Good news. I just got the code running on the PowerPC machine. 
To do that, the cpu library was skipped to avoid those bits contain sse2 or sse4 intrinsics. I also turned off the TESTs, as the machine was still trying to compile the cpu testing codes even I disabled cpu library compilation. The vecmath.h was also excluded from the library compilation list in the CMakeLists. 
I was trying to make the ppc version of vectorize.h and ran into more issues with compiler and altivec libraries, which could be detected with gcc, but not g++.  So I tried the attempts mentioned to exclude cpu code from the make list and now the program was able to run the examples on CUDA GPUs. 
I didn't compile the code in OpenCL. Because I couldn't find where the administrator placed the library or it wasn't even installed on the machine. I will further test the energy and force terms from a simulation to check if the integrity of the package was comprised without compiling the cpu platforms at all. 
Thank you very much, @peastman . I will keep you posted whether the code is still working properly and if yes, add an ppc option in the CMakeLists that will compile the code only on GPU.

**Comment by hengma1001** [comment]
I went through the installation process again. There were several things that I need to change for ppc: 
1. Use GNU environment, gcc and g++ instead of xlc or xlc++, never worked for me. 
2. Disable OPENMM_BUILD_CPU_LIB, the compilation of cpu library
3. Disable BUILD_TESTING, which will build tests using sse4.1
4. Use the same configuration as PNACL in the CMakeLists to skip all the sse compilation. I added a IBM_PPC variable to the CMakeLists.txt for PowerPC. 
5. Set up the right path for CUDA_SDK_ROOT_DIR, most of time in $CUDA_HOME/samples or $HOME/NVIDIA_CUDA-x.x_Samples
6. Do everything else like on a normal machine

This will build the OpenMM on the Nvidia GPU on the machine, skipping the troublesome compilation process on the PowerPC CPU. 
The potential energy, outputted from the stateReporter is similar to those from my local intel machine.

**Comment by hengma1001** [comment]
I went through the installation process again. There were several things that I need to change for ppc:
1. (important)Use GNU environment, newer version of gcc and g++ instead of xlc or xlc++, never worked for me. /usr/bin/cc can be old version of compiler. 
2. Disable OPENMM_BUILD_CPU_LIB, the compilation of cpu library, particularly cpupme can still be on if OpenMM finds fftw
```
 OPENMM_BUILD_PME_PLUGIN          OFF   
```
3. Disable BUILD_TESTING. Some tests will build tests using sse4.1
4. Use the same configuration as PNACL in the CMakeLists to skip all the sse compilation. I added a IBM_PPC variable to the CMakeLists.txt for PowerPC.
5. Set up the right path for CUDA_SDK_ROOT_DIR, most of time in $CUDA_HOME/samples or $HOME/NVIDIA_CUDA-x.x_Samples
6. Do everything else like on a normal machine

This will build the OpenMM on the Nvidia GPU on the machine, skipping the troublesome compilation process on the PowerPC CPU.
The potential energy, outputted from the stateReporter is similar to those from my local intel machine.

**Comment by jmjwozniak** [comment]
Does anyone have a diff for PowerPC that addresses the code changes in the steps noted on Mar 27 above?  Thanks

**Comment by hengma1001** [comment]
@jmjwozniak For that part, I defined a 'IBM_PPC' variable in the CMakeLists.txt, which was set true once a ppc architecture was detected, and changed the commands in line 227 into
```
IF (NOT (ANDROID OR PNACL OR IBM_PPC OR (WIN32 AND OPENMM_BUILD_STATIC_LIB)))
...
ENDIF (NOT (ANDROID OR PNACL OR IBM_PPC OR (WIN32 AND OPENMM_BUILD_STATIC_LIB)))
```
You can also just comment L227-L234 out. It will also work.

**Comment by jchodera** [comment]
Would either of you be able to contribute your changes as a PR? We unfortunately don't have access to Power architecture to test or develop on ourselves at the moment.

**Comment by hengma1001** [comment]
The GPU code can be compiled on the PowerPC machine, Summitdev from OLCF, with correct setup excluding the CPU part from compilation. 
I would like to contribute to the package with the Altivec version of vectorize_ppc.h and Cmake setup for PPC architecture and possibly gromacs parser. Meanwhile, I will also try to check with my advisor for @peastman about the access to those machines, which might take a while. 😅

**Comment by jchodera** [comment]
Awesome!

**Comment by mds322** [comment]
@hengma


I was able to build on Summit a few months ago (not summit-dev), but had difficulties getting the GPUs to be recognized. Did you already push your cmake changes for the summit-dev build?


I had to modify source code to get it to build on summit, so if you found a work-around using just cmake, I would be interested in taking a look

________________________________

**Comment by hengma1001** [comment]
I built the code and it can run on the login nodes in Summitdev, which has 4 P100 GPUs. But my version also had troubles with the GPU configurations on the computing nodes, the jsrun wasn't able to start OpenMM.

**Comment by jchodera** [comment]
@peastman: Did the code contributions needed to compile for Power architecture (e.g. Summit) make it in? Should we consider building a Power conda package of OpenMM for 7.3.0 (`ppc64le`)?

**Comment by peastman** [comment]
No, nothing else has happened with this.  @hengma1001 any word on it?

**Comment by hengma1001** [comment]
@peastman Did you started the process with OLCF for access to SummitDev? 
I have finished one version of vectorize_ppc.h, which passed the TestVectorize part. But when testing with CPU, there were a lot of errors, as follow. 
```
Testing CPU code of OpenMM.
../TestCpuCheckpoints
Done
../TestCpuCompoundIntegrator
Done
../TestCpuCustomGBForce
exception: Assertion failure at TestCustomGBForce.h:153.  Expected -70527.7, found -71348
../TestCpuCustomManyParticleForce
Done
../TestCpuCustomNonbondedForce
exception: Assertion failure at TestCustomNonbondedForce.h:1039.  Expected 4327.47, found 4336.32
../TestCpuDispersionPME
exception: Assertion failure at TestDispersionPME.h:1214.  Expected 12.9545, found 13.59
../TestCpuEwald
exception: Assertion failure at TestEwald.h:95.  Expected -430767, found -233416
../TestCpuGBSAOBCForce
exception: Assertion failure at TestGBSAOBCForce.h:237.  Expected 0, found 1519.9
../TestCpuGayBerneForce
Done
../TestCpuHarmonicAngleForce
Done
../TestCpuLangevinIntegrator
Done
../TestCpuNeighborList
Done
../TestCpuNonbondedForce
exception: Assertion failure at TestNonbondedForce.h:217.   Expected [0, -23.679, 0], found [-23.679, 0, 0]
../TestCpuPeriodicTorsionForce
Done
../TestCpuRBTorsionForce
Done
../TestCpuSettle
Done
```

**Comment by peastman** [comment]
What is OLCF?  I never received any information about any process to start.

**Comment by mds322** [comment]
OLCF is the Oak Ridge Leadership Computing Facility.  You should be able to get ahold of them with at https://www.olcf.ornl.gov/for-users/

The other person that was working on this (hengma) has an affiliation with them.

**Comment by hengma1001** [comment]
@peastman I will recheck with my PI about it and try to start the process asap.

**Comment by jchodera** [comment]
@hengma1001 : Thanks so much! We appreciate this!

**Comment by mds322** [comment]
If you need yet another set of eyes, I have access to summit and summit-dev as well. I did manage a build a few months back, but it refused to function at all (it wouldn't find the GPU's and the build required quite a bit of hacking at the source, which I am sure likely introduced a bunch of new bugs).

________________________________

**Comment by hengma1001** [comment]
This is the version of OpenMM that I modified for Summitdev. 
<https://github.com/hengma1001/openmm_ppc>
I updated the `CMakeLists.txt` and added `openmmapi/include/openmm/internal/vectorize_ppc.h` to the package. It can be compiled with the gcc-7.1 on summitdev so far without any errors, but the CPU code with my vectorization is still troublesome. Hopefully, it can be resolved soon.

**Comment by peastman** [comment]
This is implemented in #2212.  Could you give it a try and see if it works for you?

**Comment by hengma1001** [comment]
I just tested the ppc version on Summitdev. 
The openMM is now on PowerPC!! Congratulations, @peastman. 
Thank you for your effort. 
```
-bash-4.2$ python -m simtk.testInstallation 
There are 3 Platforms available:

1 Reference - Successfully computed forces
2 CPU - Successfully computed forces
3 CUDA - Successfully computed forces

Median difference in forces between platforms:

Reference vs. CPU: 1.98337e-06
Reference vs. CUDA: 2.15156e-05
CPU vs. CUDA: 2.14218e-05
```

One thing is that the older version (before 3.11) of cmake was not very pleased with `add_compile_definitions` function and newer version (3.12) from conda-forge didn't include the `ccmake` function. I had a few hiccups getting the `cmake ..` running. 

Other than that, everything worked like a charm.
